File: api/src/schemas/household.ts
import { z } from 'zod';
import { objectIdOrStringSchema } from './obj-id';
import { randomUUID } from 'crypto';
import { emailSchema } from './auth/auth';
import { deviceSchema, deviceSourceSchema } from './devices';
import { validateRooms } from '../util';

/**
 * Coordinates using longitude and latitude
 */
export const coordinatesSchema = z.object({
  /** Latitude */
  lat: z.number().min(-89).max(90),
  /** Longitude */
  long: z.number().min(-180).max(180),
});

/** Coordinates in the format [latitude, longitude] */
export type Coordinates = z.infer<typeof coordinatesSchema>;

/**
 * The role of a household member.
 * - owner: The owner of the household. Can do anything
 * - admin: Has elevated permissions.
 * - dweller: A regular member of the household.
 */
export const memberRoleSchema = z.enum(['owner', 'admin', 'dweller']);

/**
 * The role of a household member.
 * - owner: The owner of the household. Can do anything
 * - admin: Has elevated permissions.
 * - dweller: A regular member of the household.
 */
export type MemberRole = z.infer<typeof memberRoleSchema>;

// TODO: update the permissions to allow for more granular control such as
// setting permissions per room or per device (or exceptions at that level as well)

/**
 * A household dweller's permissions
 */
export const memberPermissionsSchema = z
  .object({
    // TODO : add per device permissions

    /** Whether the user permissions to view and control appliances */
    appliances: z.boolean(),
    /**
     * Whether the user has permissions to view other member's health devices.
     * It's important to note that users can always view their own health devices
     */
    health: z.boolean(),
    /** Whether the user has permissions to view and control security devices */
    security: z.boolean(),
    /** Whether the user has permissions to view and control energy devices */
    energy: z.boolean(),
  })
  .strict();

/** A household dweller's permissions */
export type MemberPermissions = z.infer<typeof memberPermissionsSchema>;

const _memberSchema = z.object({
  /** Member's User ID */
  id: objectIdOrStringSchema,
  /** Member's role */
  role: memberRoleSchema,
  /**
   * Member's permissions.
   *
   * This is only required if the user is a dweller
   */
  permissions: memberPermissionsSchema.optional(),
});

/** Household member schema */
export const memberSchema = _memberSchema.refine(
  ({ role, permissions }) => {
    if (role === 'dweller') {
      return !!permissions;
    }
    return true;
  },
  {
    path: ['permissions'],
    message: 'Permissions are required for dwellers',
  },
);

export type Member = z.infer<typeof memberSchema>;

// TODO: add more room types
export const householdRoomTypeSchema = z.enum([
  'living',
  'kitchen',
  'bathroom',
  'bedroom',
  'other',
]);

/** Household room schema */
export const householdRoomSchema = z.object({
  id: z.string(),
  /** Room name */
  name: z
    .string()
    .min(1, { message: 'Room name cannot be empty' }) // Prevent empty strings
    .trim(),
  /** Room type */
  type: householdRoomTypeSchema,
  /** Room floor */
  floor: z.number().int(),
  /**
   * Rooms connected to this room. This is used for laying out rooms in the app
   */
  connectedRooms: z.object({
    top: z.string().optional(),
    bottom: z.string().optional(),
    left: z.string().optional(),
    right: z.string().optional(),
  }),
});
export type HouseholdRoom = z.infer<typeof householdRoomSchema>;

export type HouseholdMember = z.infer<typeof memberSchema>;

export const householdCreateRequestDataSchema = z.object({
  /** Household name */
  name: z.string(),
  /** Household long and lat coordinates */
  coordinates: coordinatesSchema.optional(),
  /** Number of floors */
  floors: z.number().int().min(1),
  /**
   * Floors offset, for example if there's a basement then the offset would be
   * -1. If there are multiple basements then the offset would be -2, etc.
   * or if the first floor is above ground and so on.
   */
  floorsOffset: z.number().int().default(0),
  /** Household's rooms */
  rooms: z.array(householdRoomSchema).min(1),
});

export const householdInviteSchema = _memberSchema
  .extend({
    inviteId: objectIdOrStringSchema,
  })
  .refine(
    ({ role, permissions }) => {
      if (role === 'dweller' && !permissions) {
        return false;
      }
      return true;
    },
    {
      path: ['permissions'],
      message: 'Permissions are required for dweller role',
    },
  );

export type HouseholdInvite = z.infer<typeof householdInviteSchema>;

export const uiHouseholdInvite = z.object({
  inviteId: objectIdOrStringSchema,
  householdName: z.string().optional(),
  ownerName: z.string().optional(),
});

export type UIHouseholdInvite = z.infer<typeof uiHouseholdInvite>;

export type HouseholdRequestData = z.infer<
  typeof householdCreateRequestDataSchema
>;

export const householdDeviceSchema = deviceSchema.extend({
  /** The room the device is assigned to */
  roomId: z.string(),
});

export type HouseholdDevice = z.infer<typeof householdDeviceSchema>;

const defaultRoom: HouseholdRoom = {
  id: randomUUID(),
  type: 'living',
  floor: 0,
  name: 'Living Room',
  connectedRooms: {},
};

const _householdSchema = householdCreateRequestDataSchema.extend({
  _id: objectIdOrStringSchema.optional(),
  owner: objectIdOrStringSchema,
  members: z.array(memberSchema),
  invites: z.array(householdInviteSchema),
  rooms: z.array(householdRoomSchema).default([defaultRoom]),
  floors: z.number().int().min(1).max(500),
  floorsOffset: z.number().int().optional(),
  devices: z.array(householdDeviceSchema).default([]),
});

export const householdSchema = _householdSchema.superRefine((data, ctx) => {
  // room configuration
  if (!validateRooms(data.rooms)) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: 'Invalid room configuration',
      path: ['rooms'],
      fatal: true,
    });
  }

  const roomMap = new Map<string, HouseholdRoom>();

  data.rooms.forEach((room) => {
    roomMap.set(room.id, room);
    if (room.floor < 0 || room.floor >= data.floors) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Invalid room floor',
        path: ['rooms', room.id],
        fatal: true,
      });
    }
  });

  // check devices
  for (const device of data.devices) {
    if (!roomMap.has(device.roomId)) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Device is assigned to a non-existent room',
        fatal: true,
        path: ['devices', device.id],
      });
    }
  }
});

export type Household = z.infer<typeof householdSchema>;

export const householdInfoSchema = _householdSchema
  .pick({
    _id: true,
    name: true,
    owner: true,
    floors: true,
    coordinates: true,
  })
  .extend({
    /** The number of members in the household */
    members: z.number().int(),
  });

export type HouseholdInfo = z.infer<typeof householdInfoSchema>;

export const uiMemberSchema = z.object({
  id: objectIdOrStringSchema,
  name: z.string().nonempty(),
  role: memberRoleSchema,
  permissions: memberPermissionsSchema.optional(),
});

export type UIMember = z.infer<typeof uiMemberSchema>;

export const uiInvitedMember = uiMemberSchema.extend({
  inviteId: objectIdOrStringSchema,
});

export type UIInvitedMember = z.infer<typeof uiInvitedMember>;

export const uiHouseholdSchema = _householdSchema.extend({
  members: z.array(uiMemberSchema).optional(),
  invites: z.array(uiInvitedMember).optional(),
});

export type UIHousehold = z.infer<typeof uiHouseholdSchema>;

export function householdToInfo(h: Household) {
  const x: HouseholdInfo = {
    _id: h._id,
    name: h.name,
    owner: h.owner,
    floors: h.floors,
    members: h.members.length,
  };

  return householdInfoSchema.parse({ x });
}

export const roomRequestDataSchema = z.object({
  rooms: z.array(householdRoomSchema).min(1),
});
export type RoomRequestData = z.infer<typeof roomRequestDataSchema>;

const _invMemberSchema = z.object({
  householdId: objectIdOrStringSchema,
  memberId: objectIdOrStringSchema,
  email: emailSchema,
  role: memberRoleSchema,
  permissions: memberPermissionsSchema.optional(),
});

export const inviteMemberSchema = _invMemberSchema
  .omit({
    memberId: true,
    householdId: true,
  })
  .refine(
    ({ role, permissions }) => {
      if (role === 'dweller' && !permissions) {
        return false;
      }
      return true;
    },
    {
      path: ['permissions'],
      message: 'Permissions are required for dweller role',
    },
  );
export type InviteMember = z.infer<typeof inviteMemberSchema>;

export const modifyMemberSchema = _invMemberSchema
  .omit({
    householdId: true,
    email: true,
  })
  .refine(
    ({ role, permissions }) => {
      if (role === 'dweller' && !permissions) {
        return false;
      }
      return true;
    },
    {
      path: ['permissions'],
      message: 'Permissions are required for dweller role',
    },
  );

export type ModifyMember = z.infer<typeof modifyMemberSchema>;

export const respondToInviteDataSchema = z.object({
  inviteId: objectIdOrStringSchema,
  response: z.boolean(),
});

export const transferSchema = z.object({
  newOwnerId: objectIdOrStringSchema,
});

export const pairDeviceSchema = z.object({
  devices: z.array(
    z.object({
      id: z.string().nonempty(),
      source: deviceSourceSchema,
      roomId: z.string().nonempty(),
    }),
  ),
});

/* Error types */

export enum InvalidHouseholdType {
  DOES_NOT_EXIST,
  INVALID_ID,
}

export class InvalidHouseholdError extends Error {
  public readonly type: InvalidHouseholdType;
  constructor(type: InvalidHouseholdType, message?: string) {
    super(
      `Invalid Household${message ? `: ${message}` : type === InvalidHouseholdType.INVALID_ID ? 'Invalid household Id' : 'Household does not exist'}`,
    );
    this.name = 'InvalidHouseholdError';
    this.type = type;
    Object.setPrototypeOf(this, InvalidHouseholdError.prototype);
  }
}

export class MissingPermissionsError extends Error {
  constructor() {
    super('Household member permissions are missing');
    this.name = 'MissingPermissionsError';
    Object.setPrototypeOf(this, MissingPermissionsError.prototype);
  }
}

export class AlreadyMemberError extends Error {
  constructor() {
    super('User is already a member of the household');
    this.name = 'AlreadyMemberError';
    Object.setPrototypeOf(this, AlreadyMemberError.prototype);
  }
}

export class AlreadyInvitedError extends Error {
  constructor() {
    super('User is already invited to the household');
    this.name = 'AlreadyInvitedError';
    Object.setPrototypeOf(this, AlreadyInvitedError.prototype);
  }
}

export class InvalidInviteError extends Error {
  constructor() {
    super('Invalid invite');
    this.name = 'InvalidInvite';
    Object.setPrototypeOf(this, InvalidInviteError.prototype);
  }
}

export class InvalidRoomsError extends Error {
  constructor() {
    super('Invalid rooms');
    this.name = 'InvalidRoomsError';
    Object.setPrototypeOf(this, InvalidRoomsError.prototype);
  }
}

export class DeviceAlreadyPairedError extends Error {
  constructor() {
    super('Device is already paired to another household');
    this.name = 'DeviceAlreadyPairError';
    Object.setPrototypeOf(this, DeviceAlreadyPairedError.prototype);
  }
}

File: app/lib/services/household.dart
import 'package:dio/dio.dart';
import 'package:smartify/services/http/http_client.dart';

class HouseholdService {
  late final Dio _dio;
  // ignore: unused_field
  late final SmartifyHttpClient _httpClient;

  HouseholdService._(this._dio, this._httpClient);

  static Future<HouseholdService> create() async {
    final httpClient = await SmartifyHttpClient.instance;
    return HouseholdService._(httpClient.dio, httpClient);
  }
/*
export const householdSchema = householdCreateRequestDataSchema.extend({
  _id: objectIdOrStringSchema.optional(),
  owner: objectIdOrStringSchema,
  members: z.array(memberSchema),
  invites: z.array(householdInviteSchema),
  rooms: z.array(householdRoomSchema).default([defaultRoom]),
  floors: z.number().int().min(1).max(500),
  floorsOffset: z.number().int().optional(),
});
  */

  Future<Household?> createHousehold(
      String name, int floors, List<HouseholdRoom> rooms,
      [int? floorsOffset]) async {
    try {
      final body = {
        'name': name,
        'floors': floors,
        'rooms': rooms.map((r) => r.toMap()).toList(),
        if (floorsOffset != null) 'floorsOffset': floorsOffset,
      };

      final response = await _dio.post('/households/new', data: body);

      final responseBody = response.data;

      try {
        final household = Household(
          id: responseBody['_id'],
          name: responseBody['name'],
          ownerId: responseBody['owner'],
          floors: responseBody['floors'],
          floorsOffset: responseBody['floorsOffset'],
          rooms: (responseBody['rooms'] as List)
              .map((r) => HouseholdRoom(
                    id: r['id'],
                    name: r['name'],
                    type: r['type'],
                    floor: r['floor'],
                    connectedRooms: RoomConnections(
                      top: r['connectedRooms']['top'],
                      bottom: r['connectedRooms']['bottom'],
                      left: r['connectedRooms']['left'],
                      right: r['connectedRooms']['right'],
                    ),
                  ))
              .toList(),
          members: (responseBody['members'] as List)
              .map((m) => HouseholdMember(
                    id: m['_id'],
                    name: m['name'],
                    role: m['role'],
                    permissions: HouseholdPermissions(
                      appliances: m['permissions']['appliances'],
                      health: m['permissions']['health'],
                      security: m['permissions']['security'],
                      energy: m['permissions']['energy'],
                    ),
                  ))
              .toList(),
          invites: (responseBody['invites'] as List)
              .map((i) => HouseholdInvite(
                    inviteId: i['_id'],
                    userId: i['userId'],
                    householdName: i['householdName'].toString(),
                    senderName: i['senderName'].toString(),
                  ))
              .toList(),
        );
        print('Household: $household');
        return household;
      } catch (e) {
        print('Error creating household: $e');
      }
    } on DioError catch (e) {
      print('Dio Error creating household: ${e.message}');
      if (e.response != null && e.response!.data != null) {
        if (e.response!.data != null) {
          print('error response data: ${e.response!.data}');
          final error = e.response!.data as Map<String, dynamic>;
          print(
              'Error creating household: ${error['error'] ?? error['error']}');
          print('Error creating household details: ${error['details']}');
        }
      }
    } catch (e) {
      print('Error creating household: $e');
    }
    return null;
  }

  Future<List<HouseholdInfo>> getUserHouseholds() async {
    try {
      final response = await _dio.get('/households');

      return (response.data as List)
          .map((h) => HouseholdInfo(
                id: h['_id'],
                name: h['name'],
                ownerId: h['owner'],
                floors: h['floors'],
                membersCount: h['members'],
              ))
          .toList();
    } on DioError catch (e) {
      print('Dio Error getting households: ${e.message}');
      if (e.response != null && e.response!.data != null) {
        if (e.response!.data != null) {
          print('error response data: ${e.response!.data}');
          final error = e.response!.data as Map<String, dynamic>;
          print(
              'Error getting households: ${error['error'] ?? error['error']}');
          print('Error getting households details: ${error['details']}');
        }
      }
    } catch (e) {
      print('Error getting households: $e');
    }
    return [];
  }

  Future<Household?> getHousehold(String householdId) async {
    try {
      final response = await _dio.get('/households/$householdId');
      final responseBody = response.data;
      try {
        final household = Household(
          id: responseBody['_id'],
          name: responseBody['name'],
          ownerId: responseBody['owner'],
          floors: responseBody['floors'],
          floorsOffset: responseBody['floorsOffset'],
          rooms: (responseBody['rooms'] as List)
              .map((r) => HouseholdRoom(
                    id: r['id'],
                    name: r['name'],
                    type: r['type'],
                    floor: r['floor'],
                    connectedRooms: RoomConnections(
                      top: r['connectedRooms']['top'],
                      bottom: r['connectedRooms']['bottom'],
                      left: r['connectedRooms']['left'],
                      right: r['connectedRooms']['right'],
                    ),
                  ))
              .toList(),
          members: (responseBody['members'] as List)
              .map((m) => HouseholdMember(
                    id: m['_id'],
                    name: m['name'],
                    role: m['role'],
                    permissions: HouseholdPermissions(
                      appliances: m['permissions']['appliances'],
                      health: m['permissions']['health'],
                      security: m['permissions']['security'],
                      energy: m['permissions']['energy'],
                    ),
                  ))
              .toList(),
          invites: (responseBody['invites'] as List)
              .map((i) => HouseholdInvite(
                    inviteId: i['_id'],
                    userId: i['userId'],
                    householdName: i['householdName'].toString(),
                    senderName: i['senderName'].toString(),
                  ))
              .toList(),
        );
        print('Household: $household');
        return household;
      } catch (e) {
        print('Error getting household: $e');
      }
    } on DioError catch (e) {
      print('Dio Error getting household: ${e.message}');
      if (e.response != null && e.response!.data != null) {
        if (e.response!.data != null) {
          print('error response data: ${e.response!.data}');
          final error = e.response!.data as Map<String, dynamic>;
          print(
              'Error getting household: ${error['error'] ?? error['message']}');
          print('Error getting household details: ${error['details']}');

          if (error['error'] != null) {
            throw Exception(error['error']);
          }
        }
      }
    } catch (e) {
      print('Error getting household: $e');
    }
    return null;
  }

  Future<HouseholdInfo?> getHouseholdInfo(String householdId) async {
    try {
      final response = await _dio.get('/households/$householdId/info');
      final h = response.data;

      return HouseholdInfo(
        id: h['_id'],
        name: h['name'],
        ownerId: h['owner'],
        floors: h['floors'],
        membersCount: h['members'],
      );
    } on DioError catch (e) {
      print('Dio Error getting households: ${e.message}');
      if (e.response != null && e.response!.data != null) {
        if (e.response!.data != null) {
          print('error response data: ${e.response!.data}');
          final error = e.response!.data as Map<String, dynamic>;
          print(
              'Error getting households: ${error['error'] ?? error['error']}');
          print('Error getting households details: ${error['details']}');
        }
      }
    } catch (e) {
      print('Error getting households: $e');
    }
    return null;
  }

  //  Get User Invites
  Future<List<HouseholdInvite>> getUserInvites() async {
    try {
      final response = await _dio.get('/households/invites');
      return (response.data as List)
          .map((invite) => HouseholdInvite(
                inviteId: invite['_id'].toString(),
                userId: invite['userId'].toString(),
                householdName: invite['householdName'].toString(),
                senderName: invite['senderName'].toString(),
              ))
          .toList();
    } on DioError catch (e) {
      if (e.response?.statusCode == 401) {
        throw Exception('Unauthorized - Please log in');
      }
      throw Exception('Failed to fetch invites: ${e.message}');
    }
  }

  // Respond to Invite
  Future<Household> respondToInvite({
    required String inviteId,
    required bool response,
  }) async {
    try {
      final res = await _dio.post(
        '/households/invite/respond',
        data: {'inviteId': inviteId, 'response': response},
      );

      return _parseHousehold(res.data);
    } on DioError catch (e) {
      switch (e.response?.statusCode) {
        case 400:
          throw Exception(e.response?.data['error'] ?? 'Invalid invite');
        case 403:
          throw Exception('Permission denied');
        case 404:
          throw Exception('Invite not found');
        case 409:
          throw Exception('User is already a member');
        default:
          throw Exception('Failed to respond to invite');
      }
    }
  }

  // Invite Member
  Future<Household> inviteMember({
    required String householdId,
    required String email,
    required String role,
    required HouseholdPermissions permissions,
  }) async {
    try {
      final res = await _dio.post(
        '/households/$householdId/invite',
        data: {
          'email': email,
          'role': role,
          'permissions': {
            'appliances': permissions.appliances,
            'health': permissions.health,
            'security': permissions.security,
            'energy': permissions.energy,
          },
        },
      );
      return _parseHousehold(res.data);
    } on DioError catch (e) {
      switch (e.response?.statusCode) {
        case 400:
          throw Exception(e.response?.data['error'] ?? 'Invalid request');
        case 403:
          throw Exception('Only owners/admins can invite');
        case 404:
          throw Exception('Household not found');
        case 409:
          throw Exception(e.response?.data['error'] ?? 'User already invited');
        default:
          throw Exception('Failed to send invitation');
      }
    }
  }

  // Helper method to parse Household from JSON
  Household _parseHousehold(Map<String, dynamic> json) {
    return Household(
      id: json['_id'].toString(),
      name: json['name'].toString(),
      ownerId: json['owner'].toString(),
      floors: json['floors'] as int,
      floorsOffset: json['floorsOffset'] as int? ?? 0,
      rooms: (json['rooms'] as List)
          .map((r) => HouseholdRoom(
                id: r['id'].toString(),
                name: r['name'].toString(),
                type: r['type'].toString(),
                floor: r['floor'] as int,
                connectedRooms: RoomConnections(
                  top: r['connectedRooms']['top']?.toString(),
                  bottom: r['connectedRooms']['bottom']?.toString(),
                  left: r['connectedRooms']['left']?.toString(),
                  right: r['connectedRooms']['right']?.toString(),
                ),
              ))
          .toList(),
      members: (json['members'] as List)
          .map((m) => HouseholdMember(
                id: m['_id'].toString(),
                name: m['name'].toString(),
                role: m['role']?.toString(),
                permissions: m['permissions'] != null
                    ? HouseholdPermissions(
                        appliances: m['permissions']['appliances'] as bool,
                        health: m['permissions']['health'] as bool,
                        security: m['permissions']['security'] as bool,
                        energy: m['permissions']['energy'] as bool,
                      )
                    : null,
              ))
          .toList(),
      invites: (json['invites'] as List)
          .map((i) => HouseholdInvite(
                inviteId: i['_id'].toString(),
                userId: i['userId'].toString(),
                householdName: i['householdName'].toString(),
                senderName: i['senderName'].toString(),
              ))
          .toList(),
    );
  }
}

class HouseholdInfo {
  final String id;
  final String name;
  final String ownerId;
  final int floors;
  final int membersCount;
  const HouseholdInfo({
    required this.id,
    required this.name,
    required this.ownerId,
    required this.floors,
    required this.membersCount,
  });

  @override
  String toString() {
    return 'HouseholdInfo{id: $id, name: $name, ownerId: $ownerId, floors: $floors, membersCount: $membersCount}';
  }
}

class Household {
  final String id;
  final String name;
  final String ownerId;
  final int floors;
  final int floorsOffset;
  final List<HouseholdRoom> rooms;
  final List<HouseholdMember> members;
  final List<HouseholdInvite> invites;

  const Household({
    required this.id,
    required this.name,
    required this.ownerId,
    required this.floors,
    required this.floorsOffset,
    required this.rooms,
    required this.members,
    required this.invites,
  });
}

class HouseholdMember {
  final String id;
  final String name;
  final String? role;
  final HouseholdPermissions? permissions;

  const HouseholdMember(
      {required this.id, required this.name, this.role, this.permissions});
}

class HouseholdPermissions {
  final bool appliances;
  final bool health;
  final bool security;
  final bool energy;

  const HouseholdPermissions(
      {required this.health,
      required this.security,
      required this.appliances,
      required this.energy});
}

class HouseholdInvite {
  final String inviteId;
  final String userId;
  final String householdName;
  final String senderName;

  const HouseholdInvite(
      {required this.inviteId,
      required this.userId,
      required this.householdName,
      required this.senderName});
}

class HouseholdRoom {
  final String id;
  final String name;
  final String type;
  final int floor;
  final RoomConnections connectedRooms;

  const HouseholdRoom({
    required this.id,
    required this.name,
    required this.type,
    required this.floor,
    required this.connectedRooms,
  });

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'name': name,
      'type': type,
      'floor': floor,
      'connectedRooms': connectedRooms.toMap(),
    };
  }
}

class RoomConnections {
  final String? top;
  final String? bottom;
  final String? left;
  final String? right;

  const RoomConnections({
    this.top,
    this.bottom,
    this.left,
    this.right,
  });

  Map<String, dynamic> toMap() {
    return {
      if (top != null) 'top': top,
      if (bottom != null) 'bottom': bottom,
      if (left != null) 'left': left,
      if (right != null) 'right': right,
    };
  }
}
